### PR TITLE
Fixed 26721 -- Dumpdata to use ASCII-only json dump to bypass encoding errors.

### PR DESCRIPTION
**Pull request by JulienPalard**
It should do the trick, but I don't have Windows to test it.

If you have the issue and a Windows box to test it:

    pip install git+https://github.com/JulienPalard/django@mdk/dumpdata-on-windows

### CODE REVIEW

**Comment by felixxm** [comment]
Thanks for this patch, however it's not an acceptable solution. First of all we've changed JSON and YAML serializers to dump all data with Unicode by default in Django 3.1 (see ticket-29249) and we don't want to revert this change. Secondly `ensure_ascii` is recognized only by JSON serializer. Moreover this is a documentation issue so we don't expect any changes in the code.

**Comment by JulienPalard** [comment]
> First of all we've changed JSON and YAML serializers to dump all data with Unicode by default in Django 3.1

Unicode is not an encoding, while the `ensure_ascii` parameter of json.encode outputs Unicode codepoints encoded as `\uXXXX`, so I don't fully get your argument. I feel `ensure_ascii` is more Unicode-friendly than say Microsoft CP1252. To be clear, `ensure_ascii` allow to output every unicode characters by using a json representation of the codepoint:

```pycon
>>> json.dumps("☕", ensure_ascii=True)
'"\\u2615"'
```

So I don't see any drawback in using ensure_ascii here, can you be more explicit?

As per the RFC 8259, json MUST be encoded as UTF-8, currently the Django serializer encode JSON using a system-dependent encoding, so it may output valid json but may not.

> Secondly ensure_ascii is recognized only by JSON serializer.

It is ignored by the other serializers, so yes it's not nice to give superfluous ignored arguments, but it works. I do agree to try to do better, but I fear we'll derive from a simple fix that just works to a more convluted implementation.

> Moreover this is a documentation issue so we don't expect any changes in the code.

I disagree, there always two choices:

- Implement something that just works.
- Document that the implementation is wrong and provide a workaround in the doc.

I'm greatly in favor of implementing something that just work:
- It make the documentation easier to read and shorter.
- It make tools more predictable and reliable.
- Newcomers never read the whole documentation, so even if documented, they'll hit the bug and report issues, adding noise to the bug tracker.

There's two ways to fix the implementation: either use ascii-only in the json representation (single-line fix), or force UTF-8 (multi-line fix), simple is better than complicated, so why not sticking to ASCII here?

**Comment by carltongibson** [comment]
Hi @JulienPalard. Thanks for the suggestion here. 

This issue is part of a broader trend for us to insist on UTF-8 across the board. This trend has broad support from the community when discussed and isn't something we want to roll-back. 

For Windows specifically we see a number of related issues (very occasionally) come up that are **all resolved** by folks setting their system to use UTF-8. 

If folks have specific needs that can subclass and set `ensure_ascii`/`allow_unicode` as needed, but I'm not seeing at all why we should impose that across the board for a subset not opting into using UTF-8. 

> ...the Django serializer encode JSON using a system-dependent encoding...

Yes, and digging down to _Why does it have this encoding_ (here and elsewhere) can be pain to track down, but that's just us leaning on Python… — Set a sensible default and it all just works vs us needing to specify encodings at every point... — I don't think we have an appetite for that. 

I hope that makes sense. You're welcome to post the the [Developers mailing list](https://groups.google.com/g/django-developers/) if you think we're doing something terribly wrong here. 

Kind Regards,

Carlton

**Comment by pope1ni** [comment]
In addition to what @carltongibson said, Python is gradually trying to work towards making this situation better.

- Python 3.6 implemented [PEP 528](https://www.python.org/dev/peps/pep-0528/) and [PEP 529](https://www.python.org/dev/peps/pep-0529/) to change the default console and file encoding to UTF-8 on Windows.
- Python 3.7 implemented [PEP 540](https://www.python.org/dev/peps/pep-0540/) to provide a "UTF-8 Mode". Do consider whether you can enable this - it bypasses whatever the default encoding specifed by the locale on the system is and uses UTF-8.
- The upcoming Python 3.10 is implementing [PEP 597](https://www.python.org/dev/peps/pep-0597/) which will provide `EncodingWarning` that can be enabled to ensure that you are explicitly specifying the encoding to use when opening a file.

Basically there are a whole class of issues related to encoding being specified incorrectly when handling files, and typically people just want everything to work nicely. Using UTF-8 by default, in most cases, is a sensible way forward. If you need something else, you should be very explicit about what the encoding is.

**Comment by JulienPalard** [comment]
Do you mean it's easier to fix this using a stackoverflow answer than fixing it in the code?

You may be right, it's less code, implying less bugs, less maintenance, ...

I may be biased or something, but it feel wrong: for me we still have code with a known bug and known workarounds.

But as I don't use Windows I don't really care, and won't be able to really test a better patch, so... ok maybe, let's not loose more time on this.
